include/ginkgo/core/base/mpi.hpp

```diff
@@ -0,0 +1,1251 @@
+#include <tuple>
+#include <type_traits>
+
+
+#include <ginkgo/config.hpp>
+#include <ginkgo/core/base/exception.hpp>
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/polymorphic_object.hpp>
+#include <ginkgo/core/base/types.hpp>
+
+
+#if GKO_HAVE_MPI
+
+
+#include <mpi.h>
+
+
+namespace gko {
+namespace mpi {
+
+
+/*
+ * This enum is used for selecting the operation type for functions that take
+ * MPI_Op. For example the MPI_Reduce operations.
+ */
+enum class op_type {
+    sum = 1,
+    min = 2,
+    max = 3,
+    product = 4,
+    custom = 5,
+    logical_and = 6,
+    bitwise_and = 7,
+    logical_or = 8,
+    bitwise_or = 9,
+    logical_xor = 10,
+    bitwise_xor = 11,
+    max_val_and_loc = 12,
+    min_val_and_loc = 13
+};
+
+
+/*
+ * This enum specifies the threading type to be used when creating an MPI
+ * environment.
+ */
+enum class thread_type {
+    serialized = MPI_THREAD_SERIALIZED,
+    funneled = MPI_THREAD_FUNNELED,
+    single = MPI_THREAD_SINGLE,
+    multiple = MPI_THREAD_MULTIPLE
+};
+
+
+namespace detail {
+
+#define GKO_MPI_DATATYPE(BaseType, MPIType)                                  \
+    inline MPI_Datatype get_mpi_type(const BaseType&) { return MPIType; }    \
+    static_assert(true,                                                      \
+                  "This assert is used to counter the false positive extra " \
+                  "semi-colon warnings")
+
+
+GKO_MPI_DATATYPE(bool, MPI_C_BOOL);
+GKO_MPI_DATATYPE(char, MPI_CHAR);
+GKO_MPI_DATATYPE(unsigned char, MPI_UNSIGNED_CHAR);
+GKO_MPI_DATATYPE(unsigned, MPI_UNSIGNED);
+GKO_MPI_DATATYPE(int, MPI_INT);
+GKO_MPI_DATATYPE(unsigned long, MPI_UNSIGNED_LONG);
+GKO_MPI_DATATYPE(unsigned short, MPI_UNSIGNED_SHORT);
+GKO_MPI_DATATYPE(long, MPI_LONG);
+GKO_MPI_DATATYPE(float, MPI_FLOAT);
+GKO_MPI_DATATYPE(double, MPI_DOUBLE);
+GKO_MPI_DATATYPE(long double, MPI_LONG_DOUBLE);
+GKO_MPI_DATATYPE(std::complex<float>, MPI_C_COMPLEX);
+GKO_MPI_DATATYPE(std::complex<double>, MPI_C_DOUBLE_COMPLEX);
+
+
+template <typename ValueType>
+MPI_Op get_operation(gko::mpi::op_type op)
+{
+    switch (op) {
+    case gko::mpi::op_type::sum:
+        return MPI_SUM;
+    case gko::mpi::op_type::min:
+        return MPI_MIN;
+    case gko::mpi::op_type::max:
+        return MPI_MAX;
+    case gko::mpi::op_type::product:
+        return MPI_PROD;
+    case gko::mpi::op_type::logical_and:
+        return MPI_LAND;
+    case gko::mpi::op_type::bitwise_and:
+        return MPI_BAND;
+    case gko::mpi::op_type::logical_or:
+        return MPI_LOR;
+    case gko::mpi::op_type::bitwise_or:
+        return MPI_BOR;
+    case gko::mpi::op_type::logical_xor:
+        return MPI_LXOR;
+    case gko::mpi::op_type::bitwise_xor:
+        return MPI_BXOR;
+    case gko::mpi::op_type::max_val_and_loc:
+        return MPI_MAXLOC;
+    case gko::mpi::op_type::min_val_and_loc:
+        return MPI_MINLOC;
+    default:
+        GKO_NOT_SUPPORTED(op);
+    }
+}
```

Comment:
Would there be any downside to using MPI_Op directly and removing this enum type?

Reply:
You get the exception handling and a nice error message here. Additionally, I am a bit wary about mixing ginkgo defined MPI types (wrapped MPI types) and plain MPI types.

Reply:
MPI itself will also complain about non-existent operations, and this enum requires an additional conversion step from well-known MPI types. Making our library easy to use for people who are familiar with MPI sounds like a big plus to me.